src/Simplex/Chat/Store/Messages.hs

```diff
@@ -967,11 +1104,102 @@ getGroupChatBefore_ db user@User {userId} g@GroupInfo {groupId} beforeChatItemId
           |]
           (userId, groupId, search, beforeChatItemTs, beforeChatItemTs, beforeChatItemId, count)
 
-toChatItemRef :: (ChatItemId, Maybe Int64, Maybe Int64) -> Either StoreError (ChatRef, ChatItemId)
+getLocalChat :: DB.Connection -> User -> Int64 -> ChatPagination -> Maybe String -> ExceptT StoreError IO (Chat 'CTLocal)
+getLocalChat db user folderId pagination search_ = do
+  let search = fromMaybe "" search_
+  nf <- getNoteFolder db user folderId
+  case pagination of
+    CPLast count -> getLocalChatLast_ db user nf count search
+    CPAfter afterId count -> getLocalChatAfter_ db user nf afterId count search
+    CPBefore beforeId count -> getLocalChatBefore_ db user nf beforeId count search
+
+getLocalChatLast_ :: DB.Connection -> User -> NoteFolder -> Int -> String -> ExceptT StoreError IO (Chat 'CTLocal)
+getLocalChatLast_ db user nf@NoteFolder {noteFolderId} count search = do
+  let stats = ChatStats {unreadCount = 0, minUnreadItemId = 0, unreadChat = False}
+  chatItems <- getLocalChatItemsLast db user noteFolderId count search
+  pure $ Chat (LocalChat nf) (reverse chatItems) stats
+
+-- the last items in reverse order (the last item in the conversation is the first in the returned list)
+getLocalChatItemsLast :: DB.Connection -> User -> NoteFolderId -> Int -> String -> ExceptT StoreError IO [CChatItem 'CTLocal]
+getLocalChatItemsLast db User {userId} noteFolderId count search = ExceptT $ do
+  currentTs <- getCurrentTime
+  mapM (toLocalChatItem currentTs)
+    <$> DB.query
+      db
+      [sql|
+        SELECT
+          -- ChatItem
+          i.chat_item_id, i.item_ts, i.item_sent, i.item_content, i.item_text, i.item_status, i.shared_msg_id, i.item_deleted, i.item_deleted_ts, i.item_edited, i.created_at, i.updated_at, i.timed_ttl, i.timed_delete_at, i.item_live,
+          -- CIFile
+          f.file_id, f.file_name, f.file_size, f.file_path, f.file_crypto_key, f.file_crypto_nonce, f.ci_file_status, f.protocol
+        FROM chat_items i
+        LEFT JOIN files f ON f.chat_item_id = i.chat_item_id
+        WHERE i.user_id = ? AND i.note_folder_id = ? AND i.item_text LIKE '%' || ? || '%'
+        ORDER BY i.created_at DESC, i.chat_item_id DESC
+        LIMIT ?
+      |]
+      (userId, noteFolderId, search, count)
```

Comment:
no need to select all items fields - the initial approach of selecting only IDs and reusing getLocalChatItem in this commit https://github.com/simplex-chat/simplex-chat/pull/3568/commits/211a440e7892021b9c97a629baa8f7a5b01752d1 was ok. I only proposed to change filters and ORDER BY, not fully change queries. The direct chat queries were just never refactored same as groups that's why they still have full lists of fields

Reply:
Ouch... Got confused by "better to repeat logic from direct chats here".